Feature/write one d to two d array

ColmBhandal opened the pull request.

ColmBhandal committed:
Split out 2D from 1D array extension methods and added some new write methods for 2D

Array extension and 2D array extension methods & associated tests are in separate classes now
There are 2 new untested write operations available as 2D array extension methods
Some other minor tidying up done e.g. moving the Enum to its own class

ColmBhandal committed:
Fixed bugs in offset behaviour

Negative and positive offsets are now allowed
The For loop will only iterate over allowable values

ColmBhandal committed:
Unit tests for write to row and column functions for 2D array

ColmBhandal committed:
One-based array equivalents of row/column write functions & tests

ColmBhandal committed:
Added new functions to create new one based arrays from int dimensions

hctrdev commented on CsharpExtras/Extensions/ArrayExtension2D.cs:
```diff
@@ -0,0 +1,184 @@
+            if (row < 0 || row >= array.GetLength(0))
+            {
+                throw new ArgumentOutOfRangeException(string.Format(
+                    "Row index {0} is out of bounds for 2D array.", row));
```
Also include the size of the array for debugging purposes.

hctrdev commented on CsharpExtras/Extensions/ArrayExtension2D.cs:
```diff
@@ -0,0 +1,184 @@
+            if (column < 0 || column >= array.GetLength(1))
+            {
+                throw new ArgumentOutOfRangeException(string.Format(
+                    "Column index {0} is out of bounds for 2D array.", column));
```
Same as above.

hctrdev commented on CsharpExtras/Extensions/ArrayExtension2D.cs:
```diff
@@ -0,0 +1,184 @@
+            }
+        }
+
+        public static void WriteToRow<TVal>(this TVal[,] array, TVal[] rowValues, int row, int columnOffset)
```
Documentation

hctrdev commented on CsharpExtras/Extensions/ArrayExtension2D.cs:
```diff
@@ -0,0 +1,184 @@
+            }
+        }
+
+        public static void WriteToColumn<TVal>(this TVal[,] array, TVal[] columnValues, int column, int rowOffset)
```
Same as above

hctrdev commented on CsharpExtrasTest/Extensions/ArrayExtension2DTest.cs:
```diff
@@ -0,0 +1,283 @@
+            Assert.AreEqual(zipped, new string[,] { { "Hello", "world" }, { "13", "24" } });
+        }
+
+        [Test, Category("Unit")]
+        [TestCaseSource("ProviderForWrite1DArrayTo2DRow")]
+        public void Given_2DArrayOfInts_When_Write1DArrayToRow_Then_ResultIsAsExpected
+            (int[,] data, int[] dataToWrite, int row, int offset, int[] expected)
+        {
+            data.WriteToRow(dataToWrite, row, offset);            
+            for(int i = 0; i < expected.Length; i++)
+            {
+                Assert.AreEqual(expected[i], data[row, i],
+                    string.Format("Mismatch at row {0} column {1} for offset {2}", row, i, offset));
+            }
+        }
+
+        [Test, Category("Unit")]
+        [TestCaseSource("ProviderForWrite1DArrayTo2DColumn")]
+        public void Given_2DArrayOfInts_When_Write1DArrayToColumn_Then_ResultIsAsExpected
+           (int[,] data, int[] dataToWrite, int column, int offset, int[] expected)
+        {
+            data.WriteToColumn(dataToWrite, column, offset);
+            for (int i = 0; i < expected.Length; i++)
+            {
+                Assert.AreEqual(expected[i], data[i, column],
+                    string.Format("Mismatch at row {0} column {1} for offset {2}", i, column, offset));
+            }
+        }
```
Expand tests to compare the entire data array instead of just one column/row.

hctrdev commented on CsharpExtrasTest/OneBased/OneBasedArray2DTest.cs:
```diff
@@ -97,12 +99,79 @@ public void Given_1dArray_When_ConvertedTo2dArrayAcrossRows_Then_New2dArrayHasCo
+        }
+
+        #region Test Data
+        private static IEnumerable<object[]> ProviderForWrite1DArrayTo2DColumn()
+        {
+            return new List<object[]> {
+                new object[] { new int[,] { { 1, 11 }, { 2, 12 }, { 3, 13 }, { 4, 14 } }, new int[] {21, 22, 23}, 2, 1,
+                    new int[]{11, 21, 22, 23}},
+                new object[] { new int[,] { { 1, 11 }, { 2, 12 }, { 3, 13 }, { 4, 14 } }, new int[] {21, 22, 23}, 2, 0,
+                    new int[]{21, 22, 23, 14}},
+                new object[] { new int[,] { { 1, 11 }, { 2, 12 }, { 3, 13 }, { 4, 14 } }, new int[] { 1, 2, 3, 4 }, 1, 0,
+                    new int[]{1, 2, 3, 4}},
+                new object[] { new int[,] { { 1, 11 }, { 2, 12 }, { 3, 13 }, { 4, 14 } }, new int[] { 1, 2, 3, 4 }, 1, 2,
+                    new int[]{1, 2, 1, 2}},
+                new object[] { new int[,] { { 1, 11 }, { 2, 12 }, { 3, 13 }, { 4, 14 } }, new int[] { 21, 22, 23, 24 }, 1, 4,
+                    new int[]{1, 2, 3, 4}},
+                new object[] { new int[,] { { 1, 11 }, { 2, 12 }, { 3, 13 }, { 4, 14 } }, new int[] { 21, 22, 23, 24 }, 1, -1,
+                    new int[]{22, 23, 24, 4}},
+            };
+        }
+
+        private static IEnumerable<object[]> ProviderForWrite1DArrayTo2DRow()
+        {
+            return new List<object[]> {
+                new object[] { new int[,] { { 1, 2, 3, 4 }, {11, 12, 13, 14} }, new int[] {21, 22, 23}, 2, 1,
+                    new int[]{11, 21, 22, 23}},
+                new object[] { new int[,] { { 1, 2, 3, 4 }, {11, 12, 13, 14} }, new int[] {21, 22, 23}, 2, 0,
+                    new int[]{21, 22, 23, 14}},
+                new object[] { new int[,] { { 1, 2, 3, 4 }, {11, 12, 13, 14} }, new int[] { 1, 2, 3, 4 }, 1, 0,
+                    new int[]{1, 2, 3, 4}},
+                new object[] { new int[,] { { 1, 2, 3, 4 }, {11, 12, 13, 14} }, new int[] { 1, 2, 3, 4 }, 1, 2,
+                    new int[]{1, 2, 1, 2}},
+                new object[] { new int[,] { { 1, 2, 3, 4 }, {11, 12, 13, 14} }, new int[] { 21, 22, 23, 24 }, 1, 4,
+                    new int[]{1, 2, 3, 4}},
+                new object[] { new int[,] { { 1, 2, 3, 4 }, {11, 12, 13, 14} }, new int[] { 21, 22, 23, 24 }, 1, -1,
+                    new int[]{22, 23, 24, 4}},
+            };
+        }
```
Reduce the number of cases since this is already tested in the zero-based functionality

hctrdev reviewed (changes requested):
See comments

ColmBhandal committed:
Converted types to non-nullable

ColmBhandal committed:
Added more context info to exception messages in array bounds validators

ColmBhandal committed:
Documented the write to row/column functions for the array 2D extension

ColmBhandal committed:
Replaced 2D array extension tests with check for entire 2D expected array

Before, they were only checking that the column written to was correct
Now, they ensure the entire resultant array is as expected
This gives the extra protection of ensuring writes aren't happening to the wrong row/column

ColmBhandal committed:
Reduced number of test cases for one-based row/column write functions

There's no point in testing all these cases again - they're already tested in 0-based case

ColmBhandal committed:
Added a new warning to treat as an error around nullable types

Also fixed nullable annotations in existing code to get rid of that warning

ColmBhandal committed:
Added compile-time test for result storage visitor

Added this as part of a new test class which can also be used for regular tests later
